proof/refine/ARM/StateRelation.thy

```diff
@@ -252,22 +252,44 @@ definition sc_relation ::
      sc_badge sc = scBadge sc' \<and>
      sc_yield_from sc = scYieldFrom sc'"
 
+(* projection rewrite *)
 
-(* Most sched contexts should satisfy these conditions. The lemma below (valid_refills'_common_usage)
-   illustrates a common situation where these conditions hold.
+definition is_active_sc' where
+  "is_active_sc' p s' \<equiv> ((\<lambda>sc'. 0 < scRefillMax sc') |< scs_of' s') p"
 
-   It seems clear that an abbreviation of this kind is useful, but it's not clear exactly which
-   conditions it should contain. For instance, it does not contain "0 < scRefillMax sc" which one
-   may argue is equally standard.
-*)
-abbreviation (input) valid_refills' where
-  "valid_refills' sc \<equiv> scRefillMax sc \<le> length (scRefills sc) \<and> scRefillHead sc < scRefillMax sc \<and>
+lemma active_sc_at'_imp_is_active_sc':
+  "active_sc_at' scp s \<Longrightarrow> is_active_sc' scp s"
+  by (clarsimp simp: active_sc_at'_def is_active_sc'_def obj_at'_def opt_map_def projectKO_eq)
+
+lemma active_sc_at'_rewrite:
+  "active_sc_at' scp s = (is_active_sc' scp s \<and> sc_at' scp s)"
+  by (fastforce simp: active_sc_at'_def is_active_sc'_def obj_at'_def opt_map_def projectKO_eq)
+
+(* valid_refills' *)
+
+(* Most sched contexts should satisfy these conditions. These are the conditions we need for
+   the refill list circular buffer to make sense. In other words, these are the constraints
+   we expect for wrap_slice to give what we want. *)
+
+abbreviation sc_valid_refills' where
+  "sc_valid_refills' sc \<equiv> scRefillMax sc \<le> length (scRefills sc) \<and> scRefillHead sc < scRefillMax sc \<and>
                     scRefillCount sc \<le> scRefillMax sc \<and> 0 < scRefillCount sc"
 
-lemma valid_refills'_common_usage:
-  "\<exists>n. sc_relation sc n sc' \<Longrightarrow> sc_active sc \<Longrightarrow> sc_valid_refills sc \<Longrightarrow> valid_sched_context' sc' s'
-   \<Longrightarrow> valid_refills' sc'"
-  by (clarsimp simp: sc_valid_refills_def valid_sched_context'_def active_sc_def sc_relation_def)
+definition valid_refills' where
+  "valid_refills' sc_ptr s' \<equiv> (sc_valid_refills' |< scs_of' s') sc_ptr"
+
+lemma valid_refills'_nonzero_scRefillCount:
+  "valid_refills' scp s' \<Longrightarrow> ((\<lambda>sc. 0 < scRefillCount sc) |< scs_of' s') scp"
+  by (clarsimp simp: valid_refills'_def split: option.splits)
+
+lemma valid_objs'_valid_refills'':
```

Comment:
The extra `'` in the name isn't clear to me. Maybe `valid_objs'_valid_refillsE'`? Possibly as an `[elim]` rule as well?

Reply:
extra `’` is from the old version and should be removed. Thanks for catching this!